allennlp/training/exponential_moving_average.py

```diff
@@ -0,0 +1,64 @@
+from typing import Dict, Iterable
+from allennlp.models.model import Model
+
+
+class ExponentialMovingAverage:
+    """
+    Create shadow variables and maintain exponential moving average for model parameters.
+
+    Parameters
+    ----------
+    model: ``Model``, required.
+        An AllenNLP model to be optimized.
+    decay: ``float``, optional, (default=0.9999).
+        The decay rate for each step. Note that a dynamic decay rate might be used if
+        `num_updates` is provided when applying the exponential moving average.
+    """
+    def __init__(self, model: Model, decay: float = 0.9999) -> None:
+        self.decay = decay
+        self._average_values: Dict = {}
+        self._backup_values: Dict = {}
+        self._model = model
+        for name, param in model.named_parameters():
+            self._average_values[name] = param.data.clone()
+            self._backup_values[name] = param.data.clone()
+
+    def apply(self, num_updates: int = None, named_parameters: Iterable = None) -> None:
+        """
+        Apply exponential moving average to `named_parameters` if specified,
+        or we will apply this to all the trainable parameters of the model.
+
+        The optional `num_updates` parameter allows one to tweak the decay rate
+        dynamically. If passed, the actual decay rate used is:
+
+            `min(decay, (1 + num_updates) / (10 + num_updates))`
```

Comment:
where does the second equation here come from?

Reply:
You can refer to [this tensorflow implementation](https://github.com/tensorflow/tensorflow/blob/a6d8ffae097d0132989ae4688d224121ec6d8f35/tensorflow/python/training/moving_averages.py#L331).